what_url/ada_adapter.py

```diff
@@ -0,0 +1,199 @@
+from what_url._ada_wrapper import ffi, lib
+
+URL_ATTRIBUTES = (
+    'href',
+    'username',
+    'password',
+    'protocol',
+    'port',
+    'hostname',
+    'host',
+    'pathname',
+    'search',
+    'hash',
+)
+PARSE_ATTRIBUTES = URL_ATTRIBUTES + ('origin',)
+
+
+def _get_str(x):
+    ret = ffi.string(x.data, x.length).decode('utf-8') if x.length else ''
+    return ret
+
+
+def check_url(s):
+    """
+    Returns ``True`` if *s* represents a valid URL, and ``False`` otherwise.
+
+    .. code-block:: python
+
+        >>> from what_url import check_url
+        >>> check_url('bogus')
+        False
+        >>> check_url('http://a/b/c/d;p?q')
+        True
+
+    """
+    try:
+        s_bytes = s.encode('utf-8')
+    except Exception:
+        return False
+
+    ada_url = lib.ada_parse(s_bytes, len(s_bytes))
+    try:
+        return lib.ada_is_valid(ada_url)
+    finally:
+        lib.ada_free(ada_url)
+
+
+def join_url(base_url, s):
+    """
+    Return the URL that results from joining *base_url* to *s*.
+    Raises ``ValueError`` if no valid URL can be constructed.
+
+    .. code-block:: python
+
+        >>> from what_url import join_url
+        >>> base_url = 'http://a/b/c/d;p?q'
+        >>> join_url(base_url, '../g')
+        'http://a/b/g'
+
+    """
+    try:
+        base_bytes = base_url.encode('utf-8')
+        s_bytes = s.encode('utf-8')
+    except Exception:
+        raise ValueError('Invalid URL') from None
+
+    ada_url = lib.ada_parse_with_base(
+        s_bytes, len(s_bytes), base_bytes, len(base_bytes)
+    )
+    try:
+        if not lib.ada_is_valid(ada_url):
+            raise ValueError('Invalid URL') from None
+
+        return _get_str(lib.ada_get_href(ada_url))
+    finally:
+        lib.ada_free(ada_url)
+
+
+def normalize_url(s):
```

Comment:
There isn't any function named `normalize` in the WHATWG specification

Reply:
I picked this name because `os.path.normalize` is familiar to Python users.